src/check.rs

```diff
@@ -1,37 +1,66 @@
 //! Provides functionality for checking the availablility of URLs.
-use std::{fmt, path::PathBuf};
+use std::collections::HashSet;
+use std::fmt;
+use std::fs::read_to_string;
+use std::path::{Path, PathBuf};
 
 use log::debug;
+use once_cell::sync::Lazy;
+use regex::Regex;
 use url::Url;
 
+use cached::cached_key_result;
+use cached::SizedCache;
+
 use super::CheckContext;
 
+use crate::parse::parse_fragments;
+
 const PREFIX_BLACKLIST: [&str; 1] = ["https://doc.rust-lang.org"];
 
 #[derive(Debug)]
-pub enum HttpError {
-    UnexpectedStatus(Url, ureq::Response),
-    Fetch(Url, ureq::Error),
+pub enum IoError {
+    HttpUnexpectedStatus(Url, ureq::Response),
+    HttpFetch(Url, ureq::Error),
+    FileIo(String, std::io::Error),
 }
 
-impl fmt::Display for HttpError {
+impl fmt::Display for IoError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            HttpError::UnexpectedStatus(url, resp) => write!(
+            IoError::HttpUnexpectedStatus(url, resp) => write!(
                 f,
                 "Unexpected HTTP status fetching {}: {}",
                 url,
                 resp.status_text()
             ),
-            HttpError::Fetch(url, e) => write!(f, "Error fetching {}: {}", url, e),
+            IoError::HttpFetch(url, e) => write!(f, "Error fetching {}: {}", url, e),
+            IoError::FileIo(url, e) => write!(f, "Error fetching {}: {}", url, e),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Link {
+    File(String),
+    Http(Url),
```

Comment:
This should let you get rid of the `clone()`s.
```suggestion
pub enum Link<'a> {
    File(&'a str),
    Http(&'a Url),
```
If this gives you too much trouble, no problem, I can fix it in a follow-up.

Reply:
Yeah, this trickles down to `CheckError` also as `CheckError` contains `Link` via `Fragment`. Perhaps something to cleanup later?

Reply:
Sounds good, this PR is already big enough :sweat_smile: